core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java

```diff
@@ -0,0 +1,504 @@
+
+      Set<Address> joiningPartition = new HashSet<>(topology.getMembers());
+      joiningPartition.removeAll(currentPartition);
+
+      // Pick the larger partition, but if equal, always pick the partition that contains the transport's coordinator
+      if (joiningPartition.size() < currentPartition.size()) {
+         return currentPartition;
+      } else if (joiningPartition.size() > currentPartition.size()){
+         return joiningPartition;
+      } else {
+         Address coordinator = rpcManager.getTransport().getCoordinator();
+         return currentPartition.contains(coordinator) ? currentPartition : joiningPartition;
+      }
+   }
+
+   private void updateCacheEntries(CacheEntry<K,V> entry, LocalizedCacheTopology topology, PartitionMergeInfo mergeInfo) {
+      if (trace) log.tracef("Updating key %s with entry %s", entry.getKey(), entry);
+      // If no partitions, then just perform update as normal put command using the preferred entries metadata
+      if (!isPartitionMerge(topology)) {
+         if (trace) log.tracef("No partition merge has occurred, putting entry %s with metadata as normal cache operation", entry);
+         cache.put(entry.getKey(), entry.getValue(), entry.getMetadata());
+         return;
+      }
+
+      // Here we update the local entries for all NEW owners as defined in the merged hash
+      // Entries that are no longer valid, i.e. an entry stored on a node which is no longer an owner, are removed during ST
+      // TODO? Possible optimisation is restrict the update below to the new primary owner and let ST handle the rest
+
+      DistributionInfo distributionInfo = topology.getDistribution(entry.getKey());
+      Set<Address> keyOwners = new HashSet<>(distributionInfo.writeOwners());
+      if (distributionInfo.isWriteOwner()) {
+         if (trace) log.tracef("This node %s is a write owner for key %s so putting locally", rpcManager.getAddress(), entry.getKey());
+
+         keyOwners.remove(rpcManager.getAddress());
+         PutKeyValueCommand cmd = commandsFactory.buildPutKeyValueCommand(entry.getKey(), entry.getValue(), entry.getMetadata(), localFlags);
+         InvocationContext ctx = invocationContextFactory.createNonTxInvocationContext();
+         interceptorChain.invoke(ctx, cmd);
+      }
+
+      Map<Address, ReplicableCommand> cmdMap = new HashMap<>();
+      for (Address address : keyOwners) {
+         int topologyId = mergeInfo.getTopologyId(address);
+         PutKeyValueCommand cmd = commandsFactory.buildPutKeyValueCommand(entry.getKey(), entry.getValue(), entry.getMetadata(), remoteFlags);
+         cmd.setTopologyId(topologyId); // Set to max stable topology id received for destination node
+         cmdMap.put(address, cmd);
+
+         if (trace) log.tracef("Putting entry for key %s at address %s with topology id %s", entry, address, topologyId);
+      }
+
+      Map<Address, Response> responseMap = rpcManager.invokeRemotely(cmdMap, rpcOptions);
```

Comment:
It seems that all you do for an unsuccessful update is logging the failure (that's fine). Could you do the invocation asynchronously, then? Because this way the conflict resolution needs to wait for each roundtrip.